\section{Methods} \label{sec:methods}

After opening with a brief primer on hereditary stratigraphy and an overview of application in proposed ``gene-'' and ``species-level'' instrumentation (Section \ref{sec:genome-instrumentation}), discussion turns to cover debuted inferential techniques and corresponding validation experiments.
Three inferential techniques are introduced:
\begin{enumerate}
\item genealogical inference (Section \ref{sec:genealogical-inference}),
\item effective population size inference (Section \ref{sec:population-size-inference}), and
\item positive selection inference (Section \ref{sec:selection-inference}).
\end{enumerate}

The first two rely on ``species-level'' annotation.
The third, selection inference, applies ``gene-level'' instrumentation.
Section \ref{sec:software-data} closes with information on software implementation and supplementary materials.

\input{text/body/methods/genome-instrumentation.tex}

\input{text/body/methods/genealogical-inference.tex}

\input{text/body/methods/population-size-inference.tex}

\input{text/body/methods/selection-inference.tex}

\subsection{Software \& Data}
\label{sec:software-data}

Software, data, analyses, and supplementary materials are hosted via the Open Science Framework at \url{https://osf.io/xj5pn/}.

Data structures and algorithms associated with hereditary stratigraphy methodology are published in the \texttt{hstrat} Python package on PyPi and at \url{github.com/mmore500/hstrat} \citep{moreno2022hstrat}.
Recombination features and corresponding C++ implementations are on the project's roadmap.

Experiments benefited greatly from open source scientific software \citep{cock2009biopython,fortin2012deap,sand2014tqdist}.
The Alife data standard facilitated tool interoperation \citep{lalejini2019data}.
