\section{Methods} \label{sec:methods}

After opening with a brief primer on hereditary stratigraphy and an overview of its application in proposed ``gene-'' and ``species-level'' instrumentation (Section \ref{sec:genome-instrumentation}), discussion arrives at debuted inferential techniques and corresponding validation experiments.
Three inferential techniques are reported:
\begin{enumerate}
\item genealogical inference (Section \ref{sec:genealogical-inference}),
\item effective population size inference (Section \ref{sec:population-size-inference}), and
\item positive selection inference (Section \ref{sec:selection-inference}).
\end{enumerate}

The first two rely on ``species-level'' annotation.
Section \ref{sec:one-max} introduces the simple ``one-max'' problem used for validation experiments of genealogical and population size inference.
The third, selection inference, applies ``gene-level'' instrumentation.
Section \ref{sec:software-data} closes with information on software and data associated with this work.

\input{text/body/methods/genome-instrumentation.tex}

\subsection{One-Max Task}
\label{sec:one-max}

We used the one-max task for experiments to evaluate population size inference and genealogical inference.
This tasks selects over fixed-length bit string individuals for the fraction of 1's set.
Each individual was represented as a binary string of length 100.
Fitness was evaluated as the sum of the bits in the binary string, preferring individuals with more 1s over individuals with fewer.
Population size was 200 individuals, initialized randomly with sites set to 1 with probability 0.5.
In treatments where population size decreased during an evolutionary run, excess individuals were eliminated randomly.
In treatments where population size increased during an evolutionary run,
new population slots were filled through selection over the existing population.

Selection was performed using the tournament selection method.
Default tournament size was 2.
Tournament sizes 1 and 8 were also set in certain treatments.
Tournaments were performed synchronously, with all parents selected before turnover of the entire population.
Evolutionary runs lasted 100 generations.
Two-point crossover (mating) and bit-flip mutation with per-bit probability 0.05 were applied to all organisms at all generations prior to selection.
Operator choice and parameter selections were based on one-max example code from the Distributed Evolutionary Algorithms in Python package \citep{fortin2012deap}.

\input{text/body/methods/genealogical-inference.tex}

\input{text/body/methods/population-size-inference.tex}

\input{text/body/methods/selection-inference.tex}

\subsection{Software \& Data}
\label{sec:software-data}

Software, data, and analyses used in this work are hosted on GitHub at \url{github.com/mmore500/hstrat-recomb-concept/tree/notebooks/notebooks}.

Data structures and algorithms associated with hereditary stratigraphy methodology are published in the hstrat Python package, made available on PyPi and on GitHub at \url{github.com/mmore500/hstrat} under the MIT license \citep{moreno2022hstrat}.
Recombination features developed for this project, as well as corresponding C++ implementations of hereditary stratigraphy data structures and algorithms, are on the hstrat project's roadmap.

This project benefited greatly from open source scientific software, including BioPython \citep{cock2009biopython}, Distributed Evolutionary Algorithms in Python (DEAP) \citep{fortin2012deap}, Matplotlib \citep{hunter2007matplotlib}, NetworkX \citep{hagberg2008networkx}, Jupyter \citep{loizides2016jupyter}, pandas \citep{mckinney2010pandas}, SciPy \citep{pauli2020scipy}, seaborn \citep{waskom2021seaborn}, SymPy \citep{meurer2017sympy}, and tqdist \citep{sand2014tqdist}.
The Artificial Life data standard for phylogenies facilitated tool interoperation \citep{lalejini2019data}.
