\section{Results and Discussion} \label{sec:results}

This section reports validation experiments for debuted genealogical, population size, and positive selection inference techniques.
Results support their efficacy.

\subsection{Genealogical Inference}
\input{fig/species-example-replicates}
\input{fig/species-reconstruction-error}

Figure \ref{fig:species-example-replicates} compares phylogenetic trees reconstructed from species-level instrumentation to corresponding references extracted from perfectly-tracked sexual pedigrees.
For treatments with meaningful phylogentic structure --- the ``allopatry'' and ``ring'' treatments --- phylogenetic reconstruction largely succeeded in recovering the historical relationships between subpopulations.
In fact, for the ``allopatry'' treatment, inner node time points appear to more closely track the true generational time frames of speciation events (at generation 100 and 150) than the UPGMA-based pedigree distillation.

Figure \ref{fig:species-reconstruction-error} shows distributions of reconstruction error for each treatment.
Across the three treatments, all ten replicate reconstructions yielded quartet distance from reference below 0.66 (the null expectation for arbitrary trees) \citep{smith2020information}.
This confirms recovery of phylogenetic information in all three cases (exact binomial test, $p < 0.01$).

However, as expected, reconstruction quality on the bag population structure was marginal due to the lack of meaningful phylogenetic information available to reconstruct.
Performance on the ring and allopatry treatments was stronger, achieving quartet distances between reconstruction and reference of around 0.3 in the typical case.
However, junk phylogenetic structure within the reference phylogeny obscures the amount of meaningful reconstruction error.

\subsection{Population Size Inference}
\input{fig/ne-estimate-distributions}

Figure \ref{fig:ne-estimate-distributions} summarizes the distribution of effective population size estimates across replicates at the beginning, middle, and end of evolutionary runs.
Evidencing detection sensitivity, estimates differ across time points within all non-control treatments.
For the bottleneck and selection pressure treatments, which involve reversion to initial conditions, estimate distributions at the first and last time points are comparable, as expected.

Supplemental Figure \ref{fig:ne-example-replicates} shows ten-sample rolling estimates of population size for one replicate from each surveyed treatment.
All population estimates respond to underlying demographic changes, although the response to selection pressure relaxation appears weaker than responses to changes in population size.
Substantial estimate volatility appears across all cases.

Supplemental Figure \ref{fig:ne-detection-matrix} summarizes the detectability of underling effective population size changes.
Detection was performed by evaluating 95\% confidence interval overlap between rolling population size estimates at different time points
No false positives were detected.
Most true changes in effective population size were detected in at least nine out of ten replicates, except for the selection pressure treatment and for the last segment of the range expansion treatment.

\subsection{Positive Selection Inference}
\input{fig/selection-example-replicates}
Gene selection experiments introduced novel alleles with fitness advantages of 1.0 (strong selection), 0.1 (weaker selection), or 0.0 (no selection --- control).
Figure \ref{fig:selection-example-replicates} plots underlying gene copy count with instrument delay copy count estimates.
These bit counts were extracted from a single randomly-sampled population member.
For the strong selection treatment, allele frequency fixes rapidly and induces a sharp spike in the instrument values.
For the weaker selection treatment, allele frequency grows somewhat less rapidly and is sometimes delayed.
An instrument bit count spike is apparent, but it occurs with a smaller magnitude and more variable timing.

The sensitivity and specificity of positive gene selection detection was evaluated by surveying false-positive (i.e., detection selection on control replicates) and false-negative rates (i.e., non-detection of selection on fitness-advantaged replicates) across a range of detection threshold values.
Supplemental Figure \ref{fig:selection-sensitivity-specificity} plots detection outcomes across a range of detection thresholds.
Strong selection events can be unambiguously distinguished from neutral events, as well as from weaker selection events.
Weaker selection and neutral events were not entirely separable.
The middle-of-the-road detection threshold misidentified one neutral event and one weak positive event --- corresponding to a 10\% false-positive and 10\% false-negative rate.
